docs/multi-agent-applications.md

```diff
@@ -0,0 +1,333 @@
+#> Did you hear about the toothpaste scandal? They called it Colgate.
+print(result.usage())
+"""
+Usage(
+    requests=3, request_tokens=204, response_tokens=24, total_tokens=228, details=None
+)
+"""
+```
+
+_(This example is complete, it can be run "as is")_
+
+The control flow for this example is pretty simple and can be summarised as follows:
+
+```mermaid
+graph TD
+  START --> joke_agent
+  joke_agent --> joke_factory["joke_factory (tool)"]
+  joke_factory --> delegate_agent
+  delegate_agent --> joke_factory
+  joke_factory --> joke_agent
+  joke_agent --> END
+```
+
+### Agent Delegation and dependencies.
+
+The delegate agent needs to either have the same [dependencies](dependencies.md) as the calling agent, or dependencies which are a subset of the calling agent's dependencies.
+
+```python {title="agent_delegation_deps.py"}
+from dataclasses import dataclass
+
+import httpx
+
+from pydantic_ai import Agent, RunContext
+
+
+@dataclass
+class ClientAndKey:  # (1)!
+    http_client: httpx.AsyncClient
+    api_key: str
+
+
+joke_agent = Agent(
+    'openai:gpt-4o',
+    deps_type=ClientAndKey,  # (2)!
+    system_prompt=(
+        'Use the "joke_factory" tool to generate some jokes on the given subject, '
+        'then choose the best. You must return just a single joke.'
+    ),
+)
+
+
+@joke_agent.tool
+async def joke_factory(ctx: RunContext[ClientAndKey], count: int) -> list[str]:
+    r = await delegate_agent.run(
+        f'Please generate {count} jokes.',
+        deps=ctx.deps,  # (3)!
+        usage=ctx.usage,
+    )
+    return r.data
+
+
+delegate_agent = Agent(
+    'gemini-1.5-flash',
+    deps_type=ClientAndKey,  # (4)!
+    result_type=list[str],
+    system_prompt=(
+        'Use the "get_jokes" tool to get some jokes on the given subject, '
+        'then extract each joke into a list.'
+    ),
+)
+
+
+@delegate_agent.tool  # (5)!
+async def get_jokes(ctx: RunContext[ClientAndKey], count: int) -> str:
+    response = await ctx.deps.http_client.get(
+        'https://example.com',
+        params={'count': count},
+        headers={'Authorization': f'Bearer {ctx.deps.api_key}'},
+    )
+    response.raise_for_status()
+    return response.text
+
+
+async def main():
+    async with httpx.AsyncClient() as client:
+        deps = ClientAndKey(client, 'foobar')
+        result = await joke_agent.run('Tell me a joke.', deps=deps)
+        print(result.data)
+        #> Did you hear about the toothpaste scandal? They called it Colgate.
+        print(result.usage())  # (6)!
+        """
+        Usage(
+            requests=4,
+            request_tokens=310,
+            response_tokens=32,
+            total_tokens=342,
+            details=None,
+        )
+        """
+```
+
+1. Define a dataclass to hold the client and API key dependencies.
+2. Set the `deps_type` of the calling agent.
+3. Pass the dependencies to the delegate agent's run method within the tool call.
+4. Also set the `deps_type` of the delegate agent.
```

Comment:
As above, I think we should use different names for these agents both in the code snippets and notes